pmlc/dialect/abi/transforms/lower_to_abi.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "pmlc/dialect/abi/transforms/pass_detail.h"
+#include "pmlc/util/ids.h"
+
+using LLVMType = mlir::LLVM::LLVMType;
+
+namespace pmlc::dialect::abi {
+namespace {
+
+class LowerToABIPass final : public LowerToABIPassBase<LowerToABIPass> {
+public:
+  void runOnOperation() final;
+};
+
+void LowerToABIPass::runOnOperation() {
+  auto moduleOp = getOperation();
+  auto mainFunc =
+      mlir::dyn_cast_or_null<mlir::FuncOp>(moduleOp.lookupSymbol(networkMain));
+  if (!mainFunc) {
+    moduleOp.emitError(llvm::formatv("Unable to resolve \"{0}\"", networkMain));
+    signalPassFailure();
+    return;
+  }
+
+  auto mainTy = mainFunc.getType();
+  if (mainTy.getNumResults()) {
+    mainFunc.emitError(
+        llvm::formatv("Expected no outputs from \"{0}\"", networkMain));
+    signalPassFailure();
+    return;
+  }
+
+  // Create the loop.
+  auto loc = mainFunc.getLoc();
+  auto builder = mlir::OpBuilder{mainFunc};
+  auto loopOp = builder.create<abi::LoopOp>(loc);
+
+  // Initialize the loop's body by taking the main function's body.
+  loopOp.bodyRegion().takeBody(mainFunc.getBody());
+
+  // A function's blocks can end with std.return; since std.return expects its
+  // parent to be std.func, we need to replace any cloned top-level std.return
+  // operations with abi.done.
+  for (auto &op : llvm::make_early_inc_range(loopOp.bodyRegion().getOps())) {
+    auto returnOp = mlir::dyn_cast<mlir::ReturnOp>(op);
+    if (returnOp) {
+      builder.setInsertionPoint(returnOp);
+      builder.create<abi::TerminatorOp>(returnOp.getLoc());
+      returnOp.erase();
+    }
+  }
+
+  // Add an entry block for the init region.
+  auto *initEntryBlock = builder.createBlock(&loopOp.initRegion());
+
+  // Non-constant arguments will vary from invocation to invocation; they are
+  // intrinsically parameters of the body entry block.  Constant arguments do
+  // not vary; we pass them to the initializtion block, which can then pass them
+  // on to the body as needed.
+  //
+  // At this point, all of the arguments should be memrefs, except for a
+  // possible initial device argument -- which, if present, is always passed
+  // to the initialization block.
+  auto *bodyEntryBlock = loopOp.getBodyEntryBlock();
+  unsigned argIdx = 0;
+  mlir::Identifier constAttrId = builder.getIdentifier("tile.const");
```

Comment:
This seems out of place

Reply:
And also, why do we need to create an identifier? can we just use the string version directly?

Reply:
Is there another good way of distinguishing constant weight tensors?

Using an identifier just means a little less work for the compiler, since it's slightly faster to compare interned strings (although honestly, it's not a huge cost).

Reply:
You should look at the definition of the string-based `getStringAttr`, it calls `getIndentifier` under the hood. There's never a time where the compiler does more work because it's comparing strings, identifiers are always used. 

My point is that if there's no need to have an identifier, it's just more boilerplate that should be avoided.

Reply:
And to be clear, the 'out of place' comment is just that this line of code isn't used into a little bit later, so I'd move this line close to its usage. However, if we just use the string-based call, this whole line would go away.

Reply:
I did in fact trace through the code here.  It looked to me like it grounded out in an llvm::lower_bound and then a strcmp, but perhaps I missed something.  I can move the identifier creation, though.